lib/Sema/CSApply.cpp

```diff
@@ -6514,6 +6514,17 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
                                  Optional<Pattern*> typeFromPattern) {
   auto &ctx = cs.getASTContext();
 
+  // Diagnose conversions to invalid function types that couldn't be performed
+  // beforehand because of placeholders.
+  if (auto *fnTy = toType->getAs<FunctionType>()) {
+    auto contextTy = cs.getContextualType(expr);
+    if (cs.getConstraintLocator(locator)->isForContextualType() &&
+        contextTy && contextTy->hasPlaceholder()) {
+      TypeChecker::diagnoseInvalidFunctionType(fnTy, expr->getLoc(), None, dc,
+                                               None);
```

Comment:
What exactly are the implications of failing vs. not failing a coercion? Not sure I understand exactly what it means to diagnose an error while failing a coercion vs. not failing a coercion.

Reply:
@xedin Relatedly, the docs for `coerceToType` say "This operation cannot fail," but clients are clearly checking against failure of this function... is that comment outdated?

Reply:
I think it is. It's possible (albeit rare) for solver to produce a solution which couldn't be applied to AST, in such cases `ExprRewriter` should fail, if coercion is impossible than `coerceToType` should fail as well. In this case `coerceToType` would produce an error diagnostic and should return `nullptr`.

Reply:
It it was just a warning then coerceToType should have just continued.

Reply:
It might be better to let `diagnoseInvalidFunctionType` return a bool to indicate whether error was diagnosed and only fail `coerceToType` if so.

Reply:
Got it, makes sense. Yeah, you're right that to do this we'll have to have `diagnoseInvalidFunctionType` indicate failure.

Looking back at the "old" implementation, it looks like during type resolution we diagnosed these invalid types but didn't fail resolution, so presumably those invalid function types are getting applied to the AST. Do you think it makes sense to fail the application in all cases? I.e., remove the `diagnoseInvalidFunctionType` check from `TypeResolver::resolveASTFunctionType` and perform it only in `coerceToType` for all function types (not just those that had a placeholder), and fail coercion if so?

Reply:
Frankly I'm not sure exactly why it wasn't made to fail, I'd suggest you try to return `ErrorType` wrapping `fnTy` for cases when `diagnoseInvalidFunctionType` would produce an error in `resolveASTFunctionType` (so that code completion would still have access to underlying type) and return `nullptr` in `coerceToType`.

Reply:
Ok! I’ll give that a shot and see if there’s any unexpected fallout.

Reply:
🤞